packages/database/src/services/timesync.ts

```diff
@@ -0,0 +1,238 @@
+    return {
+      ...request,
+      server: this.microtime(),
+    };
+  }
+
+  #processResponse(response: TimeResponse): ProcessedResponse {
+    const current = this.microtime();
+    const latency = Math.abs(response.server - response.client) / 2;
+    const delta = response.server - current + latency;
+    return { latency, delta };
+  }
+
+  async #saveOffset(offset: Microseconds): Promise<void> {
+    this.#initial = false;
+    this.#offset = offset;
+    this.#log.debug('Timesync offset changed (us)', { offset });
+    this.#models.LocalSystemFact.set(FACT_TIME_OFFSET, offset.toString());
+  }
+
+  async #attempt(): Promise<void> {
+    const query = this.#query;
+    if (!query) return;
+
+    const { enabled, interval, jitter, samples } = (await this.#settings.get(
+      'timesync',
+    )) as TimesyncSettings;
+
+    // schedule the next one immediately so the interval is regular
+    setTimeout(
+      () =>
+        this.#attempt().catch((err) => {
+          this.#log.error(`Uncaught timesync error: ${err}`);
+        }),
+      interval * 1000,
+    );
+
+    // it's a bit wasteful to still schedule attempts that do nothing,
+    // but it means we can easily enable and disable timesync at runtime
+    if (!enabled) return;
+
+    let gap = 0;
+    const responses: ProcessedResponse[] = [];
+    for (let n = 0; n < samples; n += 1) {
+      await sleepAsync(gap);
+
+      // compute the next gap before we query, so if we error we don't immediately reloop
+      gap = Math.random() * jitter;
+
+      try {
+        // timeout is half the interval, so we guarantee that we never overrun
+        const response = await query(this.#makeRequest(), interval * 500);
+        if (!response) continue;
+
+        const processed = this.#processResponse(response);
+        gap += processed.latency;
+        responses.push(processed);
+        if (this.#initial) {
+          // when we don't have an offset at all, anything is better than nothing
+          this.#saveOffset(processed.delta).catch((err) => {
+            this.#log.error(`Timesync save error: ${err}`);
+          });
+        }
+      } catch (err) {
+        this.#log.error(`Timesync processing error: ${err}`);
+      }
+    }
+
+    if (responses.length % 2 === 0) {
+      // the first response is most likely to be an outlier, if we have to discard it
+      responses.shift();
+    }
+
+    // not enough for confidence
+    if (responses.length < 3) return;
+
+    responses.sort((a, b) => a.latency - b.latency);
+    const deltas = responses.map(({ delta }) => delta);
+
+    const medianIdx = Math.floor(deltas.length / 2);
+    const median = deltas[medianIdx]!;
+
+    const mean = deltas.reduce((acc, d) => acc + d, 0);
+    const variance =
+      deltas.map((d) => Math.pow(d - mean, 2)).reduce((acc, d) => acc + d, 0) / (deltas.length - 1);
+    const std = Math.sqrt(variance);
+
+    const inliers = deltas.filter((d) => d <= median + std || d >= median - std);
```

Comment:
I think this should be an &?
```suggestion
    const inliers = deltas.filter((d) => d <= median + std && d >= median - std);
```

FWIW I find the other order more of a "visual" representation
```suggestion
    const inliers = deltas.filter((d) => d >= median - std && d <= median + std);
```